source/val/validate_memory.cpp

```diff
@@ -1356,27 +1356,46 @@ spv_result_t ValidateAccessChain(ValidationState_t& _,
     auto index_type = _.FindDef(cur_word_instr->type_id());
     if (!index_type || spv::Op::OpTypeInt != index_type->opcode()) {
       return _.diag(SPV_ERROR_INVALID_ID, inst)
-             << "Indexes passed to " << instr_name
+             << "Indexes passed to Op" << spvOpcodeString(opcode)
              << " must be of type integer.";
     }
     switch (type_pointee->opcode()) {
       case spv::Op::OpTypeMatrix:
       case spv::Op::OpTypeVector:
       case spv::Op::OpTypeCooperativeMatrixNV:
       case spv::Op::OpTypeCooperativeMatrixKHR:
-      case spv::Op::OpTypeArray:
       case spv::Op::OpTypeRuntimeArray: {
         // In OpTypeMatrix, OpTypeVector, spv::Op::OpTypeCooperativeMatrixNV,
         // OpTypeArray, and OpTypeRuntimeArray, word 2 is the Element Type.
         type_pointee = _.FindDef(type_pointee->word(2));
         break;
       }
+      case spv::Op::OpTypeArray: {
+        // If using a descriptor array, make sure the outer array is not OOB
+        if (spv::Op::OpConstant == cur_word_instr->opcode()) {
+          const uint32_t cur_index = cur_word_instr->word(3);
+          auto array_length_id = _.FindDef(type_pointee->word(3));
+          if (spv::Op::OpConstant == array_length_id->opcode()) {
+            const uint32_t array_length = array_length_id->word(3);
+            if (cur_index >= array_length) {
+              return _.diag(SPV_ERROR_INVALID_ID, cur_word_instr)
+                     << "Index is out of bounds: Op" << spvOpcodeString(opcode)
+                     << " can not find index " << cur_index
+                     << " into the array <id> "
+                     << _.getIdName(type_pointee->id()) << " of length "
+                     << array_length << ".";
+            }
+          }
+        }
+        type_pointee = _.FindDef(type_pointee->word(2));
+        break;
+      }
       case spv::Op::OpTypeStruct: {
         // In case of structures, there is an additional constraint on the
         // index: the index must be an OpConstant.
         if (spv::Op::OpConstant != cur_word_instr->opcode()) {
           return _.diag(SPV_ERROR_INVALID_ID, cur_word_instr)
-                 << "The <id> passed to " << instr_name
+                 << "The <id> passed to Op" << spvOpcodeString(opcode)
                  << " to index into a "
                     "structure must be an OpConstant.";
         }
```

Comment:
@alan-baker working more into this, I found the line below makes it seem like it is legal to have a `-1` access chain index?

So this means if I have a 

`layout(set=0, binding=0) buffer foo { int x[3]; } var;`

These are valid?

```swift
%int_n1 = OpConstant %int -1
%int_5 = OpConstant %int 5

%ac = OpAccessChain %ptr %var %int_0 %int_n1
// or
%ac = OpAccessChain %sb_ptr %var %int_0 %int_5
```